equinox/_module.py

```diff
@@ -584,6 +584,15 @@ def __call__(cls, *args, **kwargs):
                 f"The following fields were not initialised during __init__: "
                 f"{missing_names}"
             )
+        # [Step 3.5] Prevent arrays from being marked as static
+        for field in dataclasses.fields(self):
+            if field.metadata.get("static", False):
+                if any(
+                    jtu.tree_map(
+                        is_array, jtu.tree_flatten(getattr(self, field.name))[0]
+                    )
+                ):
+                    raise ValueError("JAX Arrays cannot be marked as static!")
```

Comment:
Should we attempt to `hash` instead, do you think? And reject things that are unhashable. Also, `jtu.tree_flatten(...)[0] -> jtu.tree_leaves`

Other than that, I've been mulling this over and I think I'm not comfortable with the extent to which this is a breaking change. I think this is probably happening in quite a few places in harmless ways, and I'd like to be sure we don't needlessly break people. Can we switch this over to a `warnings.warn(..., stacklevel=...)` call instead? That will still help everyone it needs to help, without breaking things overnight.

Reply:
Ideally, if static field just means aux data to pytree, we just pass that to jax and make jax tell us if it's good with it or not. So hash is sort of a proxy, in my mind the best test here would be whatever it is that jax does that would break (like whatever tree operation actually relies on this aux stuff being properly defined) since I don't think jax checks it. Not sure what that operation is tho. 

I originally had hash, but your point before seemed valid, that having any arrays in the static pytree could have weird interactions with leaves/flattening/raveling even if there is a hash defined. That being said, if we are just raising a warning then I don't think it matters too much, and hash could be good (easy to check, and more direct of a proxy for what might actually fail if they try to treat it as aux data).

Reply:
FWIW I believe JAX has some holes here too, for similar reasons to us: they weren't careful enough to begin with, and now people are quietly depending on this treatment of things. (I don't recall the specifics though.)

No strong feelings on hashability vs arrays really, whatever you think best!

Reply:
I changed it to a warning, and I ultimately decided to leave it with leaves (as opposed to hash). My reasoning was probably the most common (incorrect) use of static arrays is to try to make parameters static/not differentiated. But you can make a validly hashable module/set of parameters, but then even if you mark it as static, you will still see gradients for the arrays. Hence, I went with checking the leaves for arrays.

Reply:
Hi there, thanks the interesting discussions. I was wondering if there is a specific reason to using `is_array`, which includes `np.ndarray`, rather than `isinstance(element, jnp.Array)` in the warning test?

https://github.com/patrick-kidger/equinox/blob/d7d2cb91dde3beee970d9f8f10fc3c9c7f2f0e39/equinox/_module.py#L581-L592

https://github.com/patrick-kidger/equinox/blob/d7d2cb91dde3beee970d9f8f10fc3c9c7f2f0e39/equinox/_filters.py#L19-L21

I'm asking for a package I develop where we rely on `Module` to handle custom PyTree declaration (not necessarily for neural networks). We sometimes rely on static fields for aux_data, and passing an `np.ndarray` as static also raises the warning.

Thanks !